### PR TITLE
fix: make sure Minestom services stop like Bukkit services

### LINKED ISSUE #1304: Minestom server doesn't shut down correctly
### Stacktrace

_No response_

### Actions to reproduce

Set up a dummy minestom server that starts and stops. The wrapper will not stop because there are some threads that aren't daemons so the wrapper just won't stop even though the application has terminated
```
        MinecraftServer server = MinecraftServer.init();
        server.start(System.getProperty("service.bind.host"), Integer.getInteger("service.bind.port"));
        try {
            Thread.sleep(10000);
        } catch (InterruptedException e) {
            throw new RuntimeException(e);
        }
        MinecraftServer.stopCleanly();
```

### CloudNet version

[31.08 10:51:24.867] INFO:
[31.08 10:51:24.868] INFO: CloudNet Blizzard 4.0.0-RC9 f6ca4c38
[31.08 10:51:24.868] INFO: Discord: <https://discord.cloudnetservice.eu/>
[31.08 10:51:24.868] INFO:
[31.08 10:51:24.869] INFO: ClusterId: deebb2f9-****-41cd-****-246e55129822
[31.08 10:51:24.869] INFO: NodeId: Node-1
[31.08 10:51:24.869] INFO: Head-NodeId: Node-1
[31.08 10:51:24.870] INFO: CPU usage: (P/S) .36/.34/100%
[31.08 10:51:24.870] INFO: Node services memory allocation (U/R/M): 1536/1536/4096 MB
[31.08 10:51:24.871] INFO: Threads: 55
[31.08 10:51:24.871] INFO: Heap usage: 42/256MB
[31.08 10:51:24.871] INFO: JVM: Amazon.com Inc. 17 (OpenJDK 64-Bit Server VM 17.0.4.1+9-LTS)
[31.08 10:51:24.871] INFO: Update Repo: CloudNetService/launchermeta, Update Branch: beta
[31.08 10:51:24.872] INFO:

### Other

A potential fix is to make all threads in the wrapper daemon threads. Should be fine because the node shouldn't care if the wrapper just stops responding

### Issue uniqueness

- [X] Yes, this issue is unique. There are no similar issues.

Edit:
Here is a thread dump of all non-daemon threads
```
"multithreadEventLoopGroup-1-1" #18 prio=10 os_prio=2 cpu=0.00ms elapsed=89.03s tid=0x000001fe547ef9c0 nid=0x5604 runnable  [0x00000089a82ff000]
   java.lang.Thread.State: RUNNABLE
        at sun.nio.ch.WEPoll.wait(java.base@17.0.4.1/Native Method)
        at sun.nio.ch.WEPollSelectorImpl.doSelect(java.base@17.0.4.1/WEPollSelectorImpl.java:111)
        at sun.nio.ch.SelectorImpl.lockAndDoSelect(java.base@17.0.4.1/SelectorImpl.java:129)
        - locked <0x00000000e0763928> (a io.netty5.channel.nio.SelectedSelectionKeySet)
        - locked <0x00000000e07638c8> (a sun.nio.ch.WEPollSelectorImpl)
        at sun.nio.ch.SelectorImpl.select(java.base@17.0.4.1/SelectorImpl.java:141)
        at io.netty5.channel.nio.SelectedSelectionKeySetSelector.select(SelectedSelectionKeySetSelector.java:62)
        at io.netty5.channel.nio.NioHandler.select(NioHandler.java:578)
        at io.netty5.channel.nio.NioHandler.run(NioHandler.java:361)
        at io.netty5.channel.SingleThreadEventLoop.runIO(SingleThreadEventLoop.java:192)
        at io.netty5.channel.SingleThreadEventLoop.run(SingleThreadEventLoop.java:176)
        at io.netty5.util.concurrent.SingleThreadEventExecutor.lambda$doStartThread$4(SingleThreadEventExecutor.java:774)
        at io.netty5.util.concurrent.SingleThreadEventExecutor$$Lambda$249/0x0000000800d8d248.run(Unknown Source)
        at io.netty5.util.internal.ThreadExecutorMap.lambda$apply$1(ThreadExecutorMap.java:68)
        at io.netty5.util.internal.ThreadExecutorMap$$Lambda$250/0x0000000800d8d468.run(Unknown Source)
        at io.netty5.util.concurrent.FastThreadLocalRunnable.run(FastThreadLocalRunnable.java:30)
        at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)

   Locked ownable synchronizers:
        - None

"Packet-Dispatcher-0" #22 prio=5 os_prio=0 cpu=0.00ms elapsed=88.94s tid=0x000001fe54b4a920 nid=0x3ed0 waiting on condition  [0x00000089a87ff000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.4.1/Native Method)
        - parking to wait for  <0x00000000e07656c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.4.1/LockSupport.java:341)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:506)
        at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@17.0.4.1/ForkJoinPool.java:3463)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@17.0.4.1/ForkJoinPool.java:3434)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:1623)
        at java.util.concurrent.LinkedBlockingQueue.take(java.base@17.0.4.1/LinkedBlockingQueue.java:435)
        at java.util.concurrent.ThreadPoolExecutor.getTask(java.base@17.0.4.1/ThreadPoolExecutor.java:1062)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.4.1/ThreadPoolExecutor.java:1122)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.4.1/ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)

   Locked ownable synchronizers:
        - None

"Packet-Dispatcher-1" #23 prio=5 os_prio=0 cpu=15.63ms elapsed=88.94s tid=0x000001fe5389dce0 nid=0x6838 waiting on condition  [0x00000089a88fe000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.4.1/Native Method)
        - parking to wait for  <0x00000000e07656c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.4.1/LockSupport.java:341)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:506)
        at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@17.0.4.1/ForkJoinPool.java:3463)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@17.0.4.1/ForkJoinPool.java:3434)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:1623)
        at java.util.concurrent.LinkedBlockingQueue.take(java.base@17.0.4.1/LinkedBlockingQueue.java:435)
        at java.util.concurrent.ThreadPoolExecutor.getTask(java.base@17.0.4.1/ThreadPoolExecutor.java:1062)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.4.1/ThreadPoolExecutor.java:1122)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.4.1/ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)

   Locked ownable synchronizers:
        - None

"Packet-Dispatcher-2" #25 prio=5 os_prio=0 cpu=0.00ms elapsed=88.37s tid=0x000001fe53888b60 nid=0x59e4 waiting on condition  [0x00000089a77ff000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.4.1/Native Method)
        - parking to wait for  <0x00000000e07656c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.4.1/LockSupport.java:341)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:506)
        at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@17.0.4.1/ForkJoinPool.java:3463)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@17.0.4.1/ForkJoinPool.java:3434)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:1623)
        at java.util.concurrent.LinkedBlockingQueue.take(java.base@17.0.4.1/LinkedBlockingQueue.java:435)
        at java.util.concurrent.ThreadPoolExecutor.getTask(java.base@17.0.4.1/ThreadPoolExecutor.java:1062)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.4.1/ThreadPoolExecutor.java:1122)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.4.1/ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)

   Locked ownable synchronizers:
        - None

"DestroyJavaVM" #27 prio=5 os_prio=0 cpu=250.00ms elapsed=88.35s tid=0x000001fe538899d0 nid=0xc4 waiting on condition  [0x0000000000000000]
   java.lang.Thread.State: RUNNABLE

   Locked ownable synchronizers:
        - None

"Packet-Dispatcher-3" #55 prio=5 os_prio=0 cpu=31.25ms elapsed=86.47s tid=0x000001fe547318a0 nid=0x3d94 waiting on condition  [0x00000089a8bff000]
   java.lang.Thread.State: WAITING (parking)
        at jdk.internal.misc.Unsafe.park(java.base@17.0.4.1/Native Method)
        - parking to wait for  <0x00000000e07656c0> (a java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject)
        at java.util.concurrent.locks.LockSupport.park(java.base@17.0.4.1/LockSupport.java:341)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionNode.block(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:506)
        at java.util.concurrent.ForkJoinPool.unmanagedBlock(java.base@17.0.4.1/ForkJoinPool.java:3463)
        at java.util.concurrent.ForkJoinPool.managedBlock(java.base@17.0.4.1/ForkJoinPool.java:3434)
        at java.util.concurrent.locks.AbstractQueuedSynchronizer$ConditionObject.await(java.base@17.0.4.1/AbstractQueuedSynchronizer.java:1623)
        at java.util.concurrent.LinkedBlockingQueue.take(java.base@17.0.4.1/LinkedBlockingQueue.java:435)
        at java.util.concurrent.ThreadPoolExecutor.getTask(java.base@17.0.4.1/ThreadPoolExecutor.java:1062)
        at java.util.concurrent.ThreadPoolExecutor.runWorker(java.base@17.0.4.1/ThreadPoolExecutor.java:1122)
        at java.util.concurrent.ThreadPoolExecutor$Worker.run(java.base@17.0.4.1/ThreadPoolExecutor.java:635)
        at java.lang.Thread.run(java.base@17.0.4.1/Thread.java:833)

   Locked ownable synchronizers:
        - None
```
The problem is that the netty threads aren't daemon threads and only shut down after `System.exit` but not after all threads created by the minestom server have exited normally.
`System.exit` is a workaround

### PR DESCRIPTION
**Pull request by GiantTreeLP**
### Motivation

Minestom is a very lightweight framework for implementing specialized Minecraft servers.  
Especially when the full implementation or behavior of vanilla Minecraft is not needed, Minestom offers a way to run services with very little resources.

One issue that arised in conjunction with CloudNet is that services implemented using Minestom would not properly stop. Minestom would shut down and the service would get unregistered from CloudNet but the processes would be kept running.

### Modification

This pull request mimics the behavior of vanilla implementations of the Minecraft Server which all call `System.exit(0);` when their process is done. This change makes sure that the JVM shuts down properly, even if other threads might block an ordinary shutdown from happening (ie. Netty and other networking related tasks).

### Result

Services using the Minestom framework now properly shut down and allow CloudNet to advance to the next lifecycle steps.

##### Other context

Fixes #1304

### CODE REVIEW

**Comment by GiantTreeLP** [comment]
As of now, this solution also works on server implementations build on hollow-cube/minestom-ce.

**Comment by 0utplay** [comment]
Can we maybe implement this as discussed here (https://discord.com/channels/713041279277203457/777861579659411466/1182967379488161872)

**Comment by GiantTreeLP** [comment]
> Can we maybe implement this as discussed here (https://discord.com/channels/713041279277203457/777861579659411466/1182967379488161872)

I'm going to update my implementation to match what we discussed back then.

**Comment by GiantTreeLP** [comment]
@0utplay I have updated the PR but can't initiate the build required.

Feel free to also add a review to the code, the changes are pretty minimal.

**Comment by 0utplay** [comment]
Thats because there are conflicts in the Wrapper class. I can resolve them later.

File: wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+
+  /**
+   * Find the last return instruction in the method.
+   * This method is useful to skip instructions in the method that come after the return statement, such as a
+   * label node or a line number node.
+   *
+   * @param method The method to search in
+   * @return The last return instruction in the method, or null if no return instruction was found
+   */
+  private static AbstractInsnNode findLastReturn(MethodNode method) {
```
Comment (0utplay) [inline wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java]: Can we please add a @NonNull annotation here (return type and param) and document the thrown NullPointerException if the passed method is null

**Review by 0utplay — CHANGES_REQUESTED** [review]

File: wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+import org.objectweb.asm.Opcodes;
+import org.objectweb.asm.Type;
+import org.objectweb.asm.tree.AbstractInsnNode;
+import org.objectweb.asm.tree.ClassNode;
+import org.objectweb.asm.tree.InsnList;
+import org.objectweb.asm.tree.InsnNode;
+import org.objectweb.asm.tree.MethodInsnNode;
+import org.objectweb.asm.tree.MethodNode;
+
+public class MinestomStopCleanlyTransformer implements Transformer {
```
Comment (0utplay) [inline wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java]: Make this class final please

File: wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
+/*
+ * Copyright 2019-2023 CloudNetService team & contributors
```
Comment (0utplay) [inline wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java]: You'll have to run spotless once. (or just change this manually to 2024)

File: wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java (hunk 1 above)
Comment (GiantTreeLP) [reply wrapper-jvm/src/main/java/eu/cloudnetservice/wrapper/transform/minestom/MinestomStopCleanlyTransformer.java]: There is no NullPointerException being thrown, I'll change the code to reflect that.

**Pull request merged**